Add OAuth

cjavilla-stripe opened the pull request:
r? @remi-stripe 


Looking for some feedback:

1) Is this PR too big?
2) Is this a decent param passing approach? (I'm about 95% confident that I understand the reason for using pointers)
3) Is stubbing out the server like this ok to test the oauth flow since oauth isnt supported by stripe-mock?


Here's a sample echo server that uses this successfully:

```go
package main
import (
  // "bytes"
  // "encoding/json"
  "github.com/foolin/goview/supports/echoview"
  "github.com/labstack/echo"
  "github.com/labstack/echo/middleware"
  "log"
  "net/http"
  "github.com/stripe/stripe-go"
  "github.com/stripe/stripe-go/oauth"
)

func main() {
  stripe.ClientID = "ca_123"
  stripe.Key = "sk_123"

  e := echo.New()
  e.Use(middleware.Logger())
  e.Use(middleware.Recover())
  e.Renderer = echoview.Default()

  e.GET("/", func(c echo.Context) error {
    return c.Render(http.StatusOK, "index", echo.Map{})
  })

  e.GET("/start", func(c echo.Context) error {
    // THEN...
    // client_id := "ca_123"
    // base := "https://connect.stripe.com"
    // path := "/oauth/authorize"
    // response_type := "code"
    // redirect_uri := "http://localhost:1323/callback"
    // query := fmt.Sprintf(
    //   "client_id=%s&response_type=%s&redirect_uri=%s",
    //   client_id,
    //   response_type,
    //   redirect_uri,
    // )
    // url := fmt.Sprintf("%s%s?%s", base, path, query)
    // NOW...
    url := oauth.AuthorizeURL(&stripe.AuthorizeURLParams{
      RedirectURI: stripe.String("http://localhost:1323/callback"),
    })
    return c.Redirect(http.StatusTemporaryRedirect, url)
  })

  e.GET("/callback", func(c echo.Context) error {
    // THEN...
    // client_secret := "sk_test_123"
    // code := c.QueryParam("code")
    // scope := c.QueryParam("scope")
    // state := c.QueryParam("state")
    // message := map[string]interface{}{
    //   "client_secret": client_secret,
    //   "grant_type": "authorization_code",
    //   "code": code,
    // }
    // bytesRepresentation, err := json.Marshal(message)
    // if err != nil {
    //   log.Fatalln(err)
    // }
    // resp, err := http.Post(
    //   "https://connect.stripe.com/oauth/token",
    //   "application/json",
    //   bytes.NewBuffer(bytesRepresentation),
    // )
    // if err != nil {
    //   log.Fatalln(err)
    // }
    // var result map[string]interface{}
    // json.NewDecoder(resp.Body).Decode(&result)
    // log.Println(result)

    token, err := oauth.New(&stripe.OAuthTokenParams{
      Code:      stripe.String(c.QueryParam("code")),
    })
    if err != nil {
      log.Fatalln(err)
    }
    return c.HTML(http.StatusOK, token.StripeUserID)
  })
  e.Logger.Fatal(e.Start(":1323"))
}
```

remi-stripe commented on oauth/client.go:
```diff
@@ -0,0 +1,83 @@
+  if stripe.BoolValue(params.Express) {
+    buf.WriteString("/express/oauth/authorize")
+  } else {
+    buf.WriteString("/oauth/authorize")
```
I don't think we use `WriteString` in the rest of the code. We use `stripe.FormatURLPath` instead: https://github.com/stripe/stripe-go/blob/master/card/client.go#31 which prepends `api.stripe.com` but we should find a way to re-use this logic. This avoids escaping issues and such, especially for the GET params, which is a new concept.

cjavilla-stripe replied:
Switched to use FormatURLPath, but hacked it a little to get the connect domain and fit with `/express` because `/express` url encodes to `%2fexpress`

cjavilla-stripe replied:
Switched to using out form module to build the URL query string from a params struct.

remi-stripe commented on oauth.go:
```diff
@@ -0,0 +1,52 @@
+  StripeLanding   *string
+  AlwaysPrompt    *bool
+  Express         *bool
+}
```
You're missing `stripe_user`. I am not a fan of implementing this, but @ob-stripe added this to .NET and so we likely will need to do the same in go for consistency
Params: https://github.com/stripe/stripe-dotnet/blob/master/src/Stripe.net/Services/OAuth/OAuthAuthorizeUrlStripeUserOptions.cs
PR: https://github.com/stripe/stripe-dotnet/pull/1542

cjavilla-stripe replied:
Will do :)

cjavilla-stripe commented on oauth.go:
```diff
@@ -0,0 +1,52 @@
+// OAuthToken is the value of the OAuthToken from OAuth flow.
+// https://stripe.com/docs/connect/oauth-reference#post-token
+type OAuthToken struct {
+  AccessToken            string          `json:"access_token"`
```
I didn't see which of these were actually deprecated in the docs. Are they officially deprecated, yet?

cjavilla-stripe replied:
Just wanted to bump this. I was looking here: https://stripe.com/docs/connect/oauth-reference#post-token-response and didn't see any of these deprecated. Thinking we should hide the `refresh_token`, `access_token`, `stripe_publishable_key` ? or mark as deprecated?

ob-stripe replied:
`refresh_token` and `access_token` are part of the OAuth specification so it would be awkward to not have them at all, but we can mark them as "deprecated" in favor of `stripe_user_id`.

We should probably be consistent with the .NET implementation: https://github.com/stripe/stripe-dotnet/blob/master/src/Stripe.net/Entities/OAuth/OAuthToken.cs

nirajjayantbolt commented:
Excited to get to use this soon! Is there a rough timeline for when this can get in?

cjavilla-stripe commented:
Hmm. tests are passing locally... Investigating

cjavilla-stripe commented:
Odd. I didn't have `golint` installed and `make test` didn't yell or anything. Once I installed `golint` it was clear what needed to be fixed. I should've looked at the Travis output before that last comment :)

brandur-stripe commented:
> Odd. I didn't have `golint` installed and `make test` didn't yell or anything.

Lint's in Make, but it's a separate step (`make lint`). If you want to run it, I'd suggest just doing a `make`, which will run `all`, and run all the checks that we've bundled in there.

brandur-stripe commented on oauth.go:
```diff
@@ -0,0 +1,149 @@
+	Params        `form:"*"`
+	AlwaysPrompt  *bool                  `form:"always_prompt"`
+	ClientID      *string                `form:"client_id"`
+	Express       *bool                  `form:"-"`
```
Ah, I see from reading the code below now. Do you think you could move this to a different "section" below the other parameters here and put a comment on top of it to something of the effect "Express is not sent as a parameter, but may modify the target URL of an authorize request".

(Just when reading this it's not obvious that it's even going to be used without reading the accompanying client code which is in a different file.)

brandur-stripe commented on stripe.go:
```diff
@@ -35,6 +35,13 @@ const (
 	APIURL string = "https://api.stripe.com"
 
+	// CONNECTURL is the URL for OAuth.
+	CONNECTURL string = "https://connect.stripe.com"
```
Haha, so admittedly this is not obvious at all. But I think `APIURL` was actually trying to follow normal naming conventions, but just ended up awkward because both parts were an acronym. (Maybe it should've been `ApiURL` instead in this case? Not sure, but whatever.)

So in this case let's go with `ConnectURL` here instead.

cjavilla-stripe replied:
When reviewing this, I don't think this is actually used anywhere and would be a convenient reference for library users, but probably not that valuable. Think I should just kill it?

brandur-stripe commented on stripe.go:
```diff
@@ -62,6 +69,9 @@ var EnableTelemetry = true
 // Key is the Stripe API key used globally in the binding.
 var Key string
 
+// ClientID is the Stripe Client ID used globally in the binding.
```
Could you elaborate on this comment a bit by saying that: all OAuth requests will use this `ClientID` by default unless the client ID specific request's parameter is overridden.

Also, I'm not 100% convinced that we need this one — it wouldn't that much worse for users to just keep `ClientID` in their own global config somewhere and just pass it in whenever they need a Stripe OAuth requests (of which there hopefully won't be that many).

cc @remi-stripe For thoughts as well.

cjavilla-stripe replied:
Updated the comment. Again, keeping this here and in params was for convenience.

brandur-stripe commented on oauth/client.go:
```diff
@@ -0,0 +1,84 @@
+	}
+	if stripe.StringValue(params.GrantType) == "" {
+		params.GrantType = stripe.String("authorization_code")
+	}
```
OOC, is there some particular rationale behind providing a parameter default here?

It may be a little more convenient I suppose, but I think our normal convention is just allowing the API request to fail the first time and signal back that `grant_type=authorization_code` should have been passed, at which point the user just corrects their implementation. It's one extra step, but they probably won't ever make the mistake again.

cjavilla-stripe replied:
Yeah it was all about convenience. Trying to make it so that users of this can pass some basics and we send sensible defaults.

brandur-stripe commented on oauth/client.go:
```diff
@@ -0,0 +1,84 @@
+	return getC().New(params)
+}
+
+// New creates the OAuthToken using a code after successful redirection back.
```
Could we use the "word style" name of the resource here? Like:

``` go
// New creates an OAuth token ...
```

brandur-stripe commented on oauth/client.go:
```diff
@@ -0,0 +1,84 @@
+}
+
+// AuthorizeURL builds the OAuth authorize URL.
+func AuthorizeURL(params *stripe.AuthorizeURLParams) string {
```
Will `AuthorizeURL` ever be needed by a user integration?

If not, my vote is that we keep it unexported (i.e. rename to `authorizeURL`). The tests will still be able to get at it because they're in the same package.

cjavilla-stripe replied:
Yeah this is a tool for users to create their AuthorizeURL that they will redirect to.

brandur-stripe replied:
Ah yes, now that I think about how OAuth works, that makes sense :)

brandur-stripe commented on oauth/client_test.go:
```diff
@@ -0,0 +1,234 @@
+	_ "github.com/stripe/stripe-go/testing"
+	"io/ioutil"
+	"net/http"
+	"testing"
```
This may be a local stylistic convention, but would you mind breaking the imports into two groups? That's generally how we've done it here.

1. Al the "built-ins" like `bytes`, `io/ioutil`, etc.
2. All the GOPATH packages like `assert "github.com/stretchr/testify/require"`, `stripe "github.com/stripe/stripe-go"`, `_ "github.com/stripe/stripe-go/testing"`.

(You can find examples in most other files in the project.)

Gofmt is smart enough to support this fine.

cjavilla-stripe replied:
Got it. Thanks for the heads up :)

cjavilla-stripe replied:
I think I got all of these cleaned up.

brandur-stripe commented on oauth/client.go:
```diff
@@ -0,0 +1,84 @@
+	"fmt"
+	stripe "github.com/stripe/stripe-go"
+	"github.com/stripe/stripe-go/form"
+	"net/http"
```
And same comment as in `client_test.go` about reorganizing imports into two groups.

cjavilla-stripe commented:
Thanks for the review and your patience, @brandur-stripe :)

cjavilla-stripe commented on client/api.go:
```diff
@@ -148,6 +149,8 @@ type API struct {
 	// LoginLinks is the client used to invoke login link related APIs.
 	LoginLinks *loginlink.Client
+	// OAuth is the client used to invoke /oauth APIs.
+	OAuth *oauth.Client
```
I believe I've initialized correctly below.

brandur-stripe replied:
Yep looks good.

cjavilla-stripe commented on form/form.go:
```diff
@@ -538,14 +538,15 @@ func (f *Values) Add(key, val string) {
 			buf.WriteByte('&')
 		}
-		buf.WriteString(url.QueryEscape(v.Key))
+		buf.WriteString(v.Key)
```
Switching to this form without an escape worked fine. I thought this might be a breaking change, which is why I initially started with the second encoding method. I think we control the keys tho so this _should_ be fine.

brandur-stripe replied:
Hm, actually now that I think about it, we probably do want to keep the escape — most keys should be safe, but there are a couple free-form areas in the API like `metadata` where we could have arbitrary input passed in.

Would you mind bringing back the `QueryEscape` and adopt the solution I suggested from stripe-ruby (replace encoded square brackets with real square brackets)? I believe that should be pretty safe regardless of use in form body or query string, and allows us to keep the API simple by having only one implementation. Sorry for the churn.

cjavilla-stripe replied:
Sounds good :)

cjavilla-stripe commented on form/form.go:
```diff
@@ -498,6 +499,21 @@ func (f *Values) Encode() string {
 
+// EncodeValues encodes only the values into “URL encoded” form
+// ("bar=baz&foo=quux") so that we can have keys like `stripe_user[email]=test`
+func (f *Values) EncodeValues() string {
```
I thought that not escaping the key would be considered a breaking change, so opted initially for a second encode method, but I think we control the keys here so not escaping the key seems fine.

cjavilla-stripe replied:
Solved it this way. seems to work well :)

cjavilla-stripe commented on oauth.go:
```diff
@@ -0,0 +1,149 @@
+)
+
+// OAuthStripeUserGender of the person who will be filling out a Stripe
+// application.  (International regulations require either male or female.)
```
Done

cjavilla-stripe commented on oauth.go:
```diff
@@ -0,0 +1,149 @@
+	Currency           *string                     `form:"currency"`
+	DobDay             uint64                      `form:"dob_day"`
+	DobMonth           uint64                      `form:"dob_month"`
+	DobYear            uint64                      `form:"dob_year"`
```
I think I got them all updated to pointers.

cjavilla-stripe commented on oauth.go:
```diff
@@ -0,0 +1,149 @@
+	Scope         *string                `form:"scope"`
+	State         *string                `form:"state"`
+	StripeLanding *string                `form:"stripe_landing"`
+	StripeUser    *OAuthStripeUserParams `form:"stripe_user"`
```
Yeah good call! I missed that on the first pass through the docs.

cjavilla-stripe commented on oauth.go:
```diff
@@ -0,0 +1,149 @@
+// OAuthTokens.
+type OAuthTokenParams struct {
+	Params       `form:"*"`
+	ClientSecret *string `form:"client_secret"`
```
✔️

cjavilla-stripe commented:
> Excited to get to use this soon! Is there a rough timeline for when this can get in?

@nirajjayantbolt  hopefully soon! sorry for the delay 😄

brandur-stripe commented on stripe.go:
```diff
@@ -35,6 +35,13 @@ const (
 	APIURL string = "https://api.stripe.com"
 
+	// ConnectURL is the URL for OAuth.
+	ConnectURL string = "https://connect.stripe.com"
```
So now that I look at this again, I think the duplicate constant between `APIURL` and `apiURL` was basically a refactoring mistake, and really doesn't help anything particularly. I think it's somewhat convenient that these URLs are exported in case users want them, so would you mind just dropping `connectURL` in favor `ConnectURL`?

brandur-stripe commented on oauth/client.go:
```diff
@@ -0,0 +1,85 @@
+	qs := &form.Values{}
+	form.AppendTo(qs, params)
+	return fmt.Sprintf(
+		"https://connect.stripe.com%s/oauth/authorize?%s",
```
And can we can change this to reuse `stripe.ConnectURL`?

brandur-stripe commented on stripe.go:
```diff
@@ -62,6 +69,11 @@ var EnableTelemetry = true
 
+// ClientID is the Stripe Client ID used by default for OAuth requests.
+// Relevant OAuth parameter types can also be initialized with a specific
+// ClientID that will take precidence over this global ClientID.
```
Minor spelling error: "precedence".

brandur-stripe commented on stripe.go:
```diff
@@ -62,6 +69,11 @@ var EnableTelemetry = true
+// ClientID is the Stripe Client ID used by default for OAuth requests.
+// Relevant OAuth parameter types can also be initialized with a specific
+// ClientID that will take precidence over this global ClientID.
+var ClientID string
```
@remi-stripe Any opinion on having a global `ClientID` versus just requiring that one be passed with parameters? I don't feel too strongly about it, but in general would prefer to avoid globals where possible, so I'd lean toward the latter.

cjavilla-stripe replied:
We chatted via slack and he agreed with you, so I've removed :)

brandur-stripe commented:
Oops, and realized I forgot to answer your original questions:

> Is this PR too big?

Definitely big-ish, but most of our OAuth implementations in other languages came in as part of one big PR, so it's not that unusual.

> Is this a decent param passing approach? (I'm about 95% confident that I understand the reason for using pointers)

Yeah, it was best to just follow the convention and use pointers, so you did the right thing.

The reason for pointer is that because variables in Go by default get their zero value (so a string defaults to `""`, a boolean defaults to `false`, etc.), without the pointer it's not possible to distinguish between a field that hasn't been set versus one that's been set explicitly to the zero value.

This doesn't make that much of a difference a lot of the time, but there are a couple places where the zero values are important. For example, we might want to explicitly set a string in the API to an empty string to empty it, or a boolean in the API to a `false` value, or an integer in the API to `0`. Before in stripe-go, this was only possible through some pretty gross workarounds, so we ended up switching to pointers everywhere.

> Is stubbing out the server like this ok to test the oauth flow since oauth isnt supported by stripe-mock?

Yeah, what you did there seems sane to me. Thanks for testing!

brandur-stripe commented on oauth.go:
```diff
@@ -0,0 +1,154 @@
+	Currency           *string `form:"currency"`
+	DOBDay             *uint64 `form:"dob_day"`
+	DOBMonth           *uint64 `form:"dob_month"`
+	DOBYear            *uint64 `form:"dob_year"`
```
And oops, actually if you wouldn't mind, these should be `*int64` actually. They can't be negative so technically `uint64` would be a better fit, but at one point we just decided to just go with `int64` everywhere by convention — in practice we'll never get to numbers high enough that we need the extra range `uint64` gets us, and using `int64` gives us some forward compatibility in case a field that currently doesn't allow negative numbers allows them in the future.

brandur-stripe commented on oauth/client_test.go:
```diff
@@ -0,0 +1,242 @@
+			Currency:           stripe.String("USD"),
+			DOBDay:             &dobDay,
+			DOBMonth:           &dobMonth,
+			DOBYear:            &dobYear,
```
And switching those to `int64` will also allow you to use the `string.Int64` helper down here.

cjavilla-stripe replied:
Sweet. I was wondering why that type didn't have a helper.

brandur-stripe commented on oauth.go:
```diff
@@ -0,0 +1,149 @@
+// https://stripe.com/docs/connect/oauth-reference#post-token
+type OAuthToken struct {
+	Error            OAuthError     `json:"error"`
+	ErrorDescription string         `json:"error_description"`
```
(And just flagging this comment once more in case it was missed originally.)

cjavilla-stripe replied:
Yeah I've been trying to work through this one... The shape of the errors from OAuth are different from the rest of the API. Our error handling / wrapping is failing to unmarshal because the key "error" in the OAuth errors is the type of error and the key "error" in the rest of the API seems to be the json object with error data.

Feels like the right approach is to make the connect backend handle the error with a different error type?

cjavilla-stripe commented:
Thanks for that explanation about our pointer use, @brandur-stripe. Remi also shared that with me and seeing those examples really drives it home. 

I think there are only two major things that need review:

- [ ] error handling - My current approach is to (1) attempt to deserialize json with the current struct object (top level error key pointing at error data, which is the common error shape), if that fails (2) attempt to deserialize with the top level Error struct type which has 2 new fields for OAuth specific errors. (https://github.com/stripe/stripe-go/pull/879/files#diff-faeee970b41138d8bdc51e7aa3601141)
- [ ] global client ID - @remi-stripe thoughts on global vs requiring that at call time? (https://github.com/stripe/stripe-go/pull/879/files#diff-ef7561d9a44d35bccb615dd5adb151be)

brandur-stripe commented:
> My current approach is to (1) attempt to deserialize json with the current struct object (top level error key pointing at error data, which is the common error shape), if that fails (2) attempt to deserialize with the top level Error struct type which has 2 new fields for OAuth specific errors.

Thanks for changing! I was thinking about it some more, and I think this is definitely closer to the path we want. Having a single unified error object for API or OAuth is how it works in stripe-dotnet already [1], and how it's being proposed for stripe-ruby as well [2].

I think it's a little janky that we have to do that two-phase fallback for deserializing an error object (although definitely not the worse thing, and the comment helps a lot). Did you consider alternative a custom `UnmarshalJSON` for the `Error` type? We might be able to have it try to detect an API error versus OAuth error, and then pick one path or the other based off of type.

Anyway, sorry that this is turning into more work than you probably expected :) The good news is that I think we're pretty close at this point.

[1] https://github.com/stripe/stripe-dotnet/blob/master/src/Stripe.net/Entities/StripeError.cs#L88-L90
[2] https://github.com/stripe/stripe-ruby/pull/811/files#diff-30e2a416d16a9b8714284f021b0a68e0R51

brandur-stripe commented:
[ ] And maybe the one thing we might still want to review is the use of default values for unspecified parameters in OAuth requests. I don't think we do this for any other object types, so my preference would be to lean against the pattern even if there may be some usability benefit.

One pitfall that I can see possibly happening is that the user doesn't have control anymore over whether the parameter is sent. Say they purposely didn't want to send a particular parameter, they now no longer have a great way to do that (they could do `stripe.String("")`, but I don't think we should make them resort to that).

cjavilla-stripe commented:
Removed the remaining defaults for unspecified values, removed the global ClientID, and took another swag at unmarshaling the error data by forking based on the BackendImplementation type.

r? @brandur-stripe

brandur-stripe commented on client/api.go:
```diff
@@ -272,8 +276,9 @@ func (a *API) Init(key string, backends *stripe.Backends) {
-	a.Orders = &order.Client{B: backends.API, Key: key}
+	a.OAuth = &oauth.Client{B: backends.Connect, Key: key}
 	a.OrderReturns = &orderreturn.Client{B: backends.API, Key: key}
+	a.Orders = &order.Client{B: backends.API, Key: key}
```
Thanks for fixing :)

brandur-stripe commented on error.go:
```diff
@@ -136,6 +136,10 @@ type Error struct {
+
+	// OAuth specific Error properties. Named OAuthError because of name conflict.
+	OAuthError            string `json:"error,omitempty"`
+	OAuthErrorDescription string `json:"error_description,omitempty"`
```
I actually like the "OAuth" prefix anyway — it makes it harder to use these outside of an OAuth context by accident (it's clear they're specifically for OAuth and nothing else).

cjavilla-stripe replied:
This was actually required because we use Error as a method on this object elsewhere.

brandur-stripe commented on oauth/client_test.go:
```diff
@@ -0,0 +1,267 @@
+			State:              stripe.String("NV"),
+			StreetAddress:      stripe.String("123 main"),
+			URL:                stripe.String("http://example.com"),
+			Zip:                stripe.String("12345"),
```
I just realized that "Zip" is actually an acronym (Zone Improvement Plan) and probably should've been "ZIP" instead. Don't change it though because we've already got the "Zip" convention in some fields elsewhere.

cjavilla-stripe replied:
Got it. Yeah tried to follow convention :)

brandur-stripe commented on stripe.go:
```diff
@@ -450,9 +457,20 @@ func (s *BackendImplementation) Do(req *http.Request, body *bytes.Buffer, v inte
 	var raw rawError
-	if err := s.UnmarshalJSONVerbose(res.StatusCode, resBody, &raw); err != nil {
-		return err
+	if s.Type == ConnectBackend {
```
Yeah, I like this new deserialization logic better — feels cleaner. Thanks for changing.

brandur-stripe commented on stripe.go:
```diff
@@ -450,9 +457,20 @@ func (s *BackendImplementation) Do(req *http.Request, body *bytes.Buffer, v inte
 
 // ResponseToError converts a stripe response to an Error.
 func (s *BackendImplementation) ResponseToError(res *http.Response, resBody []byte) error {
+
```
Ultra minor, but would you mind taking out this extraneously added newline?

cjavilla-stripe replied:
Done.

brandur-stripe commented:
@cjavilla-stripe New changes look great! Would you mind squashing this down? After that I think we're pretty much good to go.

@remi-stripe This looks ready to me. Do you want to take a pass for yourself before we go to release?

brandur-stripe commented on oauth/client_test.go:
```diff
@@ -0,0 +1,267 @@
+	assert.Contains(t, url, "stripe_user[zip]=12345")
+}
+
+// RoundTripFunc.
```
Can you either unexport this or put together a more complete comment?

brandur-stripe commented on oauth/client_test.go:
```diff
@@ -0,0 +1,267 @@
+// RoundTripFunc.
+type RoundTripFunc func(req *http.Request) *http.Response
+
+// RoundTrip.
```
And same with this one.

cjavilla-stripe replied:
Unexported.

brandur-stripe commented on oauth/client_test.go:
```diff
@@ -0,0 +1,267 @@
+func TestNewOAuthToken(t *testing.T) {
+	stripe.Key = "sk_123"
+
+	// stripe-mock doesn't support connect URL's so this stubs out the server.
```
Maybe consider changing all instances of "URL's" to "URLs"? I Googled and there's [no clear answer for the correct way to handle these](https://english.stackexchange.com/a/921) (sources vary), but my feeling is that the apostrophe-less version is generally preferred.

cjavilla-stripe replied:
Updated these instances.

cjavilla-stripe commented:
Thanks for the thorough review, @brandur-stripe! :)

brandur-stripe commented:
Excellent!! Thank you again CJ!

brandur-stripe commented:
Released as 61.23.0.

cjavilla-stripe commented:
🎉
